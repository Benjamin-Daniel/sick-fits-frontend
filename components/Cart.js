import React from 'react';
import { Query, Mutation } from 'react-apollo';
import gql from 'graphql-tag';
import { adopt } from 'react-adopt';
import User from './User';
import CartItem from './CartItem';
import CartStyles from './styles/CartStyles';
import Supreme from './styles/Supreme';
import CloseButton from './styles/CloseButton';
import SickButton from './styles/SickButton';
import CalcTotalPrice from '../lib/calcTotalPrice';
import FormatMoney from '../lib/formatMoney';

const LOCAL_STATE_QUERY = gql`
    query {
        cartOpen @client
    }
`;

const TOGGLE_CART_MUTAION = gql`
    mutation {
        toggleCart @client
    }
`;

const Composed = adopt({
    User:({render}) => <User>{render}</User>,
    toggleCart: ({render}) => <Mutation mutation={TOGGLE_CART_MUTAION}>{render}</Mutation>,
    localState: ({render}) => <Query query={LOCAL_STATE_QUERY}>{render}</Query>,
});

const Cart = () => (
    <Composed>
        {({ User, toggleCart, localState }) => {
            const me = User.data.me;
            if (!me) return null;
            return (
                <CartStyles open={localState.data.cartOpen}>
                    <header>
                        <CloseButton onClick={toggleCart} title='close'>&times;</CloseButton>
                        <Supreme>{me.name}'s Cart</Supreme>
                        <p>You have {me.cart.length} Item{me.cart.length === 1 ? '' : 's'} in your cart.</p>
                    </header>
                    <ul>
                        {me.cart.map(cartItem => <CartItem cartItem={cartItem} key={cartItem.id} />)}
                    </ul>
                    <footer>
                        <p>{FormatMoney(CalcTotalPrice(me.cart))}</p>
                        <SickButton>CheckOut</SickButton>
                    </footer>
                </CartStyles>
            )
        }}
    </Composed>
)

export default Cart
export { LOCAL_STATE_QUERY, TOGGLE_CART_MUTAION };