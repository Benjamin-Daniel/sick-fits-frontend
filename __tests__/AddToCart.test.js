import { mount } from "enzyme";
import wait from "waait";
import toJson from "enzyme-to-json";
import { MockedProvider } from "react-apollo/test-utils";
import { ApolloConsumer } from "react-apollo";
import AddToCart, { ADD_TO_CART_MUTAION } from "../components/AddToCart";
import { CURRENT_USER_QUERY } from "../components/User";
import { fakeCartItem, fakeUser } from "../lib/testUtils";

const mocks = [
  {
    request: { query: CURRENT_USER_QUERY },
    result: {
      data: {
        me: {
          ...fakeUser(),
          cart: []
        }
      }
    }
  },
  {
    request: { query: CURRENT_USER_QUERY },
    result: {
      data: {
        me: {
          ...fakeUser(),
          cart: [fakeCartItem()]
        }
      }
    }
  },
  {
    request: { query: ADD_TO_CART_MUTAION, variables: { id: "abc123" } },
    result: {
      data: {
        addToCart: {
          ...fakeCartItem(),
          quantity: 1
        }
      }
    }
  }
];

describe("<Cart />", () => {
  it("renders and matches snap shot", async () => {
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <AddToCart id="abc123" />
      </MockedProvider>
    );
    await wait();
    wrapper.update();
    expect(toJson(wrapper.find("button"))).toMatchSnapshot();
    // expect(wrapper.find("CartItem")).toHaveLength(1);
  });
  it("adds an item when clicked", async () => {
    let apolloClient;
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <ApolloConsumer>
          {client => {
            apolloClient = client;
            return <AddToCart id="abc123" />;
          }}
        </ApolloConsumer>
      </MockedProvider>
    );
    await wait();
    wrapper.update();
    const {
      data: { me }
    } = await apolloClient.query({ query: CURRENT_USER_QUERY });
    // expect me.cart to have length of 0
    expect(me.cart).toHaveLength(0);
    wrapper.find("button").simulate("click");
    await wait();
    // check if the item is in the cart;
    const res = await apolloClient.query({ query: CURRENT_USER_QUERY });
    const me2 = res.data.me;
    expect(me2.cart).toHaveLength(1);
    expect(me2.cart[0].id).toBe("omg123");
    expect(me2.cart[0].quantity).toBe(3);
  });

  it("changes from add to adding when clicked", async () => {
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <AddToCart id="abc123" />;
      </MockedProvider>
    );
    await wait();
    wrapper.update();
    expect(wrapper.text()).toContain("Add To Cart");
    wrapper.find("button").simulate("click");
    expect(wrapper.text()).toContain("Adding To Cart");
  });
});
