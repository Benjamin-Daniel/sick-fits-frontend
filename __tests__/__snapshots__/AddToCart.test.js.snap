// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Cart /> renders and matches snap shot 1`] = `
<button
  disabled={false}
  onClick={[Function]}
>
  Add
   To Cart
</button>
`;
